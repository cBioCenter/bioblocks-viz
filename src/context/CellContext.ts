import * as React from 'react';

import { CELL_TYPE } from '../data/chell-data';

export interface ICellContext {
  addCells: (cells: CELL_TYPE[]) => void;
  currentCells: CELL_TYPE[];
  removeAllCells: (cells: CELL_TYPE[]) => void;
  removeCells: (cells: CELL_TYPE[]) => void;
}

export const initialCellContext: ICellContext = {
  addCells: (cells: CELL_TYPE[]) => {
    return;
  },
  currentCells: [] as CELL_TYPE[],
  removeAllCells: (cells: CELL_TYPE[]) => {
    return;
  },
  removeCells: (cells: CELL_TYPE[]) => {
    return;
  },
};

const CellContext = React.createContext({
  ...initialCellContext,
});

export default CellContext;
export { CellContext };
