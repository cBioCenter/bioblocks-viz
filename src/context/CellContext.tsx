import * as React from 'react';

import { CELL_TYPE } from '../data/chell-data';

export const initialCellContext = {
  addCells: (cells: CELL_TYPE[]) => {
    return;
  },
  currentCells: new Array<CELL_TYPE>(),
  removeAllCells: () => {
    return;
  },
  removeCells: (cells: CELL_TYPE[]) => {
    return;
  },
};

export type ICellContext = typeof initialCellContext;

export class CellContextHandler extends React.Component<any, ICellContext> {
  public constructor(props: any) {
    super(props);
    this.state = {
      addCells: this.onAddCells,
      currentCells: new Array<CELL_TYPE>(),
      removeAllCells: this.onRemoveAllCells,
      removeCells: this.onRemoveCells,
    };
  }

  public render() {
    return <CellContext.Provider value={this.state}>{this.props.children}</CellContext.Provider>;
  }

  protected onAddCells = (cells: CELL_TYPE[]) => {
    this.setState({
      currentCells: cells,
    });
  };

  protected onRemoveAllCells = () => {
    this.setState({
      currentCells: [],
    });
  };

  protected onRemoveCells = (cellsToRemove: CELL_TYPE[]) => {
    const { currentCells } = this.state;
    this.setState({
      currentCells: currentCells.filter(cell => cellsToRemove.indexOf(cell) === -1),
    });
  };
}

const CellContext = React.createContext(initialCellContext);

export default CellContext;
export { CellContext };
