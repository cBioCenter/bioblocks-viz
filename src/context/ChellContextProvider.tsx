import * as React from 'react';
import { RouteComponentProps } from 'react-router';

import {
  CellContextProvider,
  ResidueContextProvider,
  SecondaryStructureContextProvider,
  SpringContextProvider,
} from '~chell-viz~/context';

export interface IChellContextProps extends Partial<RouteComponentProps> {}

/**
 * Shorthand for passing all Chell contexts down.
 *
 * @export
 * @extends {React.Component<any, any>}
 */
export class ChellContextProvider extends React.Component<IChellContextProps, any> {
  constructor(props: IChellContextProps) {
    super(props);
  }

  public render() {
    const params = new URLSearchParams(this.props.location ? this.props.location.search : '');
    // tslint:disable-next-line:no-backbone-get-set-outside-model
    const datasetLocation = params.get('name');

    return (
      <SecondaryStructureContextProvider>
        <SpringContextProvider datasetLocation={datasetLocation ? datasetLocation : ''}>
          <CellContextProvider>
            <ResidueContextProvider>{this.props.children}</ResidueContextProvider>
          </CellContextProvider>
        </SpringContextProvider>
      </SecondaryStructureContextProvider>
    );
  }
}
