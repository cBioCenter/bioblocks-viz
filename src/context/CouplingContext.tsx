import * as React from 'react';

import {
  ResidueContext,
  ResidueContextProvider,
  SecondaryStructureContext,
  SecondaryStructureContextProvider,
} from '~chell-viz~/context';
import { CouplingContainer } from '~chell-viz~/data';

export const initialCouplingContext = {
  couplingScores: new CouplingContainer(),
};

export type ICouplingContext = typeof initialCouplingContext;

export const CouplingContext = React.createContext(initialCouplingContext);

/**
 * Shorthand for passing contexts relevant for Coupling Scores - Primarily interaction with residues and secondary structures.
 *
 * @export
 * @extends {React.Component<any, ICouplingContext>}
 */
export class CouplingContextProvider extends React.Component<any, ICouplingContext> {
  constructor(props: any) {
    super(props);
    this.state = {
      couplingScores: new CouplingContainer(),
    };
  }

  public render() {
    return (
      <SecondaryStructureContextProvider>
        <ResidueContextProvider>
          <CouplingContext.Provider value={this.state}>{this.props.children}</CouplingContext.Provider>
        </ResidueContextProvider>
      </SecondaryStructureContextProvider>
    );
  }
}

export const withCouplingContext = (Component: any) => (props: any) => (
  <CouplingContext.Consumer>
    {couplingContext => (
      <SecondaryStructureContext.Consumer>
        {secStructContext => (
          <ResidueContext.Consumer>
            {residueContext => (
              <Component
                {...props}
                residueContext={{ ...residueContext }}
                secondaryStructureContext={{ ...secStructContext }}
                couplingContext={{ ...couplingContext }}
              />
            )}
          </ResidueContext.Consumer>
        )}
      </SecondaryStructureContext.Consumer>
    )}
  </CouplingContext.Consumer>
);
