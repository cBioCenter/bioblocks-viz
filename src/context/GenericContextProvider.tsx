// tslint:disable:max-classes-per-file
import * as React from 'react';

export interface IChellContext<R extends object, W extends object, RW = R & W> {
  Consumers: {
    Both: React.Consumer<RW>;
    Read: React.Consumer<R>;
    Write: React.Consumer<W>;
  };
  Providers: {
    Both: React.Provider<RW>;
    Read: React.Provider<R>;
    Write: React.Provider<W>;
  };
  allConsumers: Array<React.Consumer<R> | React.Consumer<W> | React.Consumer<RW>>;
  allProviders: Array<React.Provider<R> | React.Provider<W> | React.Provider<RW>>;
  allProvidersJSX(state: any, children: React.ReactNode): JSX.Element | React.ReactNode;
}

export const createGenericContext = <R extends object, W extends object>(
  readValue: R,
  writeValue: W,
): IChellContext<R, W> => {
  const readContext = React.createContext(readValue);
  const writeContext = React.createContext(writeValue);
  // tslint:disable-next-line:no-object-literal-type-assertion
  const bothContext = React.createContext({ ...(readValue as object), ...(writeValue as object) } as R & W);

  const Consumers = {
    Both: bothContext.Consumer,
    Read: readContext.Consumer,
    Write: writeContext.Consumer,
  };

  const Providers = {
    Both: bothContext.Provider,
    Read: readContext.Provider,
    Write: writeContext.Provider,
  };

  const allConsumers = [...Object.values(Consumers)];
  const allProviders = [...Object.values(Providers)];

  return {
    Consumers,
    Providers,
    allConsumers,
    allProviders,
    allProvidersJSX: (state, children) =>
      allProviders.reduce((prev, Curr) => <Curr value={state}>{prev}</Curr>, children),
  };
};
