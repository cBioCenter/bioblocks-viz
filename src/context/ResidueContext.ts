import * as React from 'react';
import { RESIDUE_TYPE } from '../data/chell-data';

export type ResidueSelection = Map<string, RESIDUE_TYPE[]>;

export interface IResidueContext {
  addCandidateResidues: (residues: RESIDUE_TYPE[]) => void;
  addHoveredResidues: (residues: RESIDUE_TYPE[]) => void;
  addLockedResiduePair: (residues: RESIDUE_TYPE[]) => void;
  candidateResidues: RESIDUE_TYPE[];
  clearAllResidues: () => void;
  hoveredResidues: RESIDUE_TYPE[];
  lockedResiduePairs: ResidueSelection;
  removeAllLockedResiduePairs: () => void;
  removeCandidateResidues: () => void;
  removeHoveredResidues: () => void;
  removeLockedResiduePair: (residues: RESIDUE_TYPE[]) => void;
  removeNonLockedResidues: () => void;
  toggleLockedResiduePair: (residues: RESIDUE_TYPE[]) => void;
}

export const initialResidueContext = {
  addCandidateResidues: (residues: RESIDUE_TYPE[]) => {
    return;
  },
  addHoveredResidues: (residues: RESIDUE_TYPE[]) => {
    return;
  },
  addLockedResiduePair: (residues: RESIDUE_TYPE[]) => {
    return;
  },
  candidateResidues: [] as number[],
  clearAllResidues: () => {
    return;
  },
  hoveredResidues: [] as number[],
  lockedResiduePairs: new Map(),
  removeAllLockedResiduePairs: () => {
    return;
  },
  removeCandidateResidues: () => {
    return;
  },
  removeHoveredResidues: () => {
    return;
  },
  removeLockedResiduePair: (residues: RESIDUE_TYPE[]) => {
    return;
  },
  removeNonLockedResidues: (residues: RESIDUE_TYPE[]) => {
    return;
  },
  toggleLockedResiduePair: (residues: RESIDUE_TYPE[]) => {
    return;
  },
} as IResidueContext;

const ResidueContext = React.createContext({
  ...initialResidueContext,
});

export default ResidueContext;
export { ResidueContext };
