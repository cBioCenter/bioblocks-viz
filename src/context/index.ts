export * from './CellContext';
export * from './ChellContext';
export * from './CouplingContext';
export * from './ResidueContext';
export * from './SecondaryStructureContext';
export * from './SpringContext';
