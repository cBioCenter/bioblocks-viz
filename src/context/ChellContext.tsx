import * as React from 'react';

import { CELL_TYPE, RESIDUE_TYPE } from '../data/chell-data';
import CellContext, { initialCellContext } from './CellContext';
import ResidueContext, { initialResidueContext } from './ResidueContext';
import { initialSecondaryStructureContext, SecondaryStructureContextHandler } from './SecondaryStructureContext';
import { initialSpringContext, SpringContextHandler } from './SpringContext';

export const initialState = {
  cellContext: {
    ...initialCellContext,
  },
  residueContext: {
    ...initialResidueContext,
  },
  secondaryStructureContext: {
    ...initialSecondaryStructureContext,
  },
  springContext: {
    ...initialSpringContext,
  },
};

export type ChellContextState = Readonly<typeof initialState>;

export default class ChellContext extends React.Component<any, ChellContextState> {
  public readonly state: ChellContextState = initialState;

  constructor(props: any) {
    super(props);
    this.state = {
      ...this.state,
      cellContext: {
        ...this.state.cellContext,
        addCells: this.onAddCells,
        removeAllCells: this.onRemoveAllCells,
        removeCells: this.onRemoveCells,
      },
      residueContext: {
        ...this.state.residueContext,
        addCandidateResidues: this.onAddCandidateResidues,
        addHoveredResidues: this.onAddHoveredResidues,
        addLockedResiduePair: this.onAddLockedResiduePair,
        clearAllResidues: this.onClearAllResidues,
        removeAllLockedResiduePairs: this.onRemoveAllLockedResiduePairs,
        removeCandidateResidues: this.onRemoveCandidateResidue,
        removeHoveredResidues: this.onRemoveHoveredResidues,
        removeLockedResiduePair: this.onRemoveLockedResiduePair,
        removeNonLockedResidues: this.onRemoveNonLockedResidues,
        toggleLockedResiduePair: this.onToggleLockedResiduePair,
      },
      secondaryStructureContext: {
        ...this.state.secondaryStructureContext,
      },
      springContext: {
        ...this.state.springContext,
      },
    };
  }

  public render() {
    return (
      <SecondaryStructureContextHandler>
        <SpringContextHandler>
          <CellContext.Provider value={this.state.cellContext}>
            <ResidueContext.Provider value={this.state.residueContext}>{this.props.children}</ResidueContext.Provider>
          </CellContext.Provider>
        </SpringContextHandler>
      </SecondaryStructureContextHandler>
    );
  }

  public onAddCells = (cells: CELL_TYPE[]) => {
    this.setState({
      cellContext: {
        ...this.state.cellContext,
        currentCells: cells,
      },
    });
  };

  public onRemoveAllCells = () => {
    this.setState({
      cellContext: {
        ...this.state.cellContext,
        currentCells: [],
      },
    });
  };

  public onRemoveCells = (cellsToRemove: CELL_TYPE[]) => {
    const { currentCells } = this.state.cellContext;
    this.setState({
      cellContext: {
        ...this.state.cellContext,
        currentCells: currentCells.filter(cell => cellsToRemove.indexOf(cell) === -1),
      },
    });
  };

  public onAddCandidateResidues = (candidateResidues: RESIDUE_TYPE[]) => {
    this.setState({
      residueContext: {
        ...this.state.residueContext,
        candidateResidues: candidateResidues.sort(),
      },
    });
  };

  public onAddHoveredResidues = (hoveredResidues: RESIDUE_TYPE[]) => {
    this.setState({
      residueContext: {
        ...this.state.residueContext,
        hoveredResidues: hoveredResidues.sort(),
      },
    });
  };

  public onAddLockedResiduePair = (residues: RESIDUE_TYPE[]) => {
    const { lockedResiduePairs } = this.state.residueContext;
    const sortedResidues = residues.sort();
    const residuePairKey = sortedResidues.toString();
    if (!lockedResiduePairs.has(residuePairKey)) {
      const result = new Map(lockedResiduePairs);
      result.set(residuePairKey, sortedResidues);
      this.setState({
        residueContext: {
          ...this.state.residueContext,
          lockedResiduePairs: result,
        },
      });
    }
  };

  public onClearAllResidues = () => {
    this.setState({
      residueContext: {
        ...this.state.residueContext,
        candidateResidues: [],
        hoveredResidues: [],
        lockedResiduePairs: new Map(),
      },
    });
  };

  public onRemoveAllLockedResiduePairs = () => {
    this.setState({
      residueContext: {
        ...this.state.residueContext,
        lockedResiduePairs: new Map(),
      },
    });
  };

  public onRemoveCandidateResidue = () => {
    this.setState({
      residueContext: {
        ...this.state.residueContext,
        candidateResidues: [],
      },
    });
  };

  public onRemoveHoveredResidues = () => {
    this.setState({
      residueContext: {
        ...this.state.residueContext,
        hoveredResidues: [],
      },
    });
  };

  public onRemoveLockedResiduePair = (residues: RESIDUE_TYPE[]) => {
    const residuePairKey = residues.sort().join(',');
    const { lockedResiduePairs } = this.state.residueContext;
    if (lockedResiduePairs.has(residuePairKey)) {
      const result = new Map(lockedResiduePairs);
      result.delete(residuePairKey);
      this.setState({
        residueContext: {
          ...this.state.residueContext,
          lockedResiduePairs: result,
        },
      });
    }
  };

  public onRemoveNonLockedResidues = () => {
    this.setState({
      residueContext: {
        ...this.state.residueContext,
        candidateResidues: [],
        hoveredResidues: [],
      },
    });
  };

  public onToggleLockedResiduePair = (residues: RESIDUE_TYPE[]) => {
    const { lockedResiduePairs } = this.state.residueContext;
    const sortedResidues = residues.sort();
    const residuePairKey = sortedResidues.toString();
    const result = new Map(lockedResiduePairs);
    if (!lockedResiduePairs.has(residuePairKey)) {
      result.set(residuePairKey, sortedResidues);
    } else {
      result.delete(residuePairKey);
    }
    this.setState({
      residueContext: {
        ...this.state.residueContext,
        lockedResiduePairs: result,
      },
    });
  };
}

export { ChellContext };
