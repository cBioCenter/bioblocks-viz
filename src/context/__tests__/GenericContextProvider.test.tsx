import { mount } from 'enzyme';
import * as React from 'react';

import { createGenericContext } from '~chell-viz~/context';

describe('Generic Context Provider', () => {
  it('Should render read, write, and both providers at once.', () => {
    const readContext = {
      name: 'Lucas',
    };
    const writeContext = {
      updateName: () => {
        return;
      },
    };

    const context = createGenericContext(readContext, writeContext);

    expect(context.allProvidersJSX(readContext, <></>)).toMatchSnapshot();
  });

  it('Should allow updates.', () => {
    const initialReadValue = {
      name: 'Lucas',
    };
    const initialWriteValue = {
      updateName: (name: string) => {
        return;
      },
    };

    const SmashContext = createGenericContext(initialReadValue, initialWriteValue);

    class MockContextProvider extends React.Component<any, typeof initialWriteValue & typeof initialReadValue> {
      constructor(props: any) {
        super(props);
        this.state = {
          name: 'Lucas',
          updateName: this.onUpdateName,
        };
      }

      public render() {
        return SmashContext.allProvidersJSX(this.state, this.props.children);
      }

      public onUpdateName = (name: string) => {
        this.setState({
          name,
        });
      };
    }

    // tslint:disable-next-line:max-classes-per-file
    class MockContextConsumer extends React.Component<any, any> {
      constructor(props: any) {
        super(props);
      }

      public render() {
        return (
          <SmashContext.Consumers.Read>
            {context => <div>{`Hi! My name is ${context.name}`}</div>}
          </SmashContext.Consumers.Read>
        );
      }
    }

    const wrapper = mount(
      <MockContextProvider>
        <MockContextConsumer />
      </MockContextProvider>,
    );

    expect(wrapper).toMatchSnapshot();

    (wrapper.instance() as MockContextProvider).onUpdateName('Isaac');

    wrapper.update();
    expect(wrapper).toMatchSnapshot();
  });
});
