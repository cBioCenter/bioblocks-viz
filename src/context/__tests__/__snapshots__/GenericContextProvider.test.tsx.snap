// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generic Context Provider Should allow updates. 1`] = `
<MockContextProvider>
  <MockContextConsumer>
    <div>
      Hi! My name is Lucas
    </div>
  </MockContextConsumer>
</MockContextProvider>
`;

exports[`Generic Context Provider Should allow updates. 2`] = `
<MockContextProvider>
  <MockContextConsumer>
    <div>
      Hi! My name is Isaac
    </div>
  </MockContextConsumer>
</MockContextProvider>
`;

exports[`Generic Context Provider Should render read, write, and both providers at once. 1`] = `
<Context.Provider
  value={
    Object {
      "name": "Lucas",
    }
  }
>
  <Context.Provider
    value={
      Object {
        "name": "Lucas",
      }
    }
  >
    <Context.Provider
      value={
        Object {
          "name": "Lucas",
        }
      }
    >
      <React.Fragment />
    </Context.Provider>
  </Context.Provider>
</Context.Provider>
`;
